spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java

```diff
@@ -0,0 +1,871 @@
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.DeleteSchemaUtil;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.orc.ORC;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.spark.JobGroupInfo;
+import org.apache.spark.api.java.function.MapPartitionsFunction;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoder;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BaseCopyTableSparkAction extends BaseSparkAction<CopyTable> implements CopyTable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BaseCopyTableSparkAction.class);
+  private static final String DATA_FILE_LIST_DIR = "data-file-list-to-move";
+  private static final String METADATA_FILE_LIST_DIR = "metadata-file-list-to-move";
+
+  private final Table table;
+  private final Set<String> metadataFilesToMove = Collections.synchronizedSet(Sets.newHashSet());
+  private final Set<String> manifestFilePaths = Collections.synchronizedSet(Sets.newHashSet());
+  private final Set<ManifestFile> manifestFilesToRewrite =
+      Collections.synchronizedSet(Sets.newHashSet());
+  private String dataFileListPath = null;
+  private String metadataFileListPath = null;
+  private final boolean enabledPME;
+
+  private String sourcePrefix = "";
+  private String targetPrefix = "";
+  private String startVersion = "";
+  private String endVersion = "";
+  private String stagingDir = "";
+  private Table targetTable = null;
+
+  private Table startStaticTable = null;
+  private Table endStaticTable = null;
+
+  public BaseCopyTableSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    enabledPME = table.properties().containsKey("parquet.encryption.footer.key");
+  }
+
+  @Override
+  public CopyTable rewriteLocationPrefix(String sPrefix, String tPrefix) {
+    Preconditions.checkArgument(
+        sPrefix != null && !sPrefix.isEmpty(), "Source prefix('%s') cannot be empty.", sPrefix);
+    this.sourcePrefix = sPrefix;
+
+    if (tPrefix != null) {
+      this.targetPrefix = tPrefix;
+    }
+    return this;
+  }
+
+  @Override
+  public CopyTable lastCopiedVersion(String sVersion) {
```

Comment:
it's not clear what `sVersion` refers to, so why not `newStartVersion`? Same for all the other params